core/src/main/scala/chisel3/internal/package.scala

```diff
@@ -101,4 +102,41 @@ package object internal {
     */
   private[chisel3] val ViewParent =
     Module.do_apply(new ViewParentAPI)(UnlocatableSourceInfo)
+
+  private[chisel3] def requireHasProbeTypeModifier(
+    probe:        Data,
+    errorMessage: String = ""
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Unit = {
+    val msg = if (errorMessage.isEmpty) s"Expected a probe." else errorMessage
+    if (!hasProbeTypeModifier(probe)) Builder.error(msg)
+  }
+
+  private[chisel3] def requireNoProbeTypeModifier(
+    probe:        Data,
+    errorMessage: String = ""
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Unit = {
+    val msg = if (errorMessage.isEmpty) s"Did not expect a probe." else errorMessage
+    if (hasProbeTypeModifier(probe)) Builder.error(msg)
+  }
+
+  private[chisel3] def requireHasWritableProbeTypeModifier(
+    probe:        Data,
+    errorMessage: String = ""
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Unit = {
+    val msg = if (errorMessage.isEmpty) s"Expected a writable probe." else errorMessage
+    requireHasProbeTypeModifier(probe)
```

Comment:
Should `errorMessage` or `msg` be propagated to `requireHasProbeTypeModifier` here?